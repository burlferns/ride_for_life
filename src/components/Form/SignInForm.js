import React from 'react';
import styled from "styled-components";
import { useFormik } from 'formik';
import * as Yup from 'yup';

import InputComp from './InputComp.js';
import DropDownSelectComp from './DropDownSelectComp.js';

import iconEnvelope from '../../icons/fontawesome/envelope.svg';
import iconLock from '../../icons/fontawesome/lock.svg';

const StylForm = styled.form`
  display:grid;
  grid-template-columns: 100%;
  grid-template-rows: auto;
  grid-template-areas: "email" "pass" "select";
`;

const EmailInput = styled(InputComp)`
  grid-area: email;
`;

const PassInput = styled(InputComp)`
  grid-area: pass;
`;

const StylSelect = styled(DropDownSelectComp)`
  grid-area: select;

  margin-bottom:30px;
`;

export default function(props) {
  const className = props.className;

  const formik = useFormik({
    initialValues: {
      email: '',
      passwd: '',
      userType: ''
    },
    validationSchema: Yup.object({
      email: Yup.string()
        .email('Please input a valid email address')
        .required('Email address is required'),
      passwd: Yup.string()
        .required("Please input a password"),
      userType: Yup.string()
        .oneOf(["mom", "driver"],"Please choose account type")
        .required("Please choose account type")
    }),
    onSubmit: function(values) {
      console.log('onSubmit func, arguments=',arguments);
      console.log('onSubmit func, values=',values);
    }
  })


  return (
    <StylForm onSubmit={formik.handleSubmit} className={className}>

      <EmailInput
        name='email'
        description='Email Address'
        type='text'
        icon={iconEnvelope}
        error={formik.touched.email && formik.errors.email}
        onChange={formik.handleChange}
        onBlur={formik.handleBlur}
        value={formik.values.email}
      />

      <PassInput
        name='passwd'
        description='Password'
        type='password'
        icon={iconLock}
        error={formik.touched.passwd && formik.errors.passwd}
        onChange={formik.handleChange}
        onBlur={formik.handleBlur}
        value={formik.values.passwd}
      />

      <StylSelect 
        name='userType'
        description='Account Type'
        error={formik.touched.userType && formik.errors.userType}
        onChange={formik.handleChange} 
        onBlur={formik.handleBlur}
        options={['Mom', 'Driver']}
        
        
      />


      <button type="submit" style={{width:'100px', margin:'30px'}}>Submit</button>

      



    </StylForm>    
  )
}

