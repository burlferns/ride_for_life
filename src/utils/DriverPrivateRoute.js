import React from 'react';
import { Route, Redirect } from "react-router-dom";
import {useSelector} from 'react-redux';

const selectorFunc = state=>state.userData.userType;

export default function(props) {
  const routeAttributes = props.routeAttributes;
  const element = props.element;
  const userType = useSelector(selectorFunc);

  return (
    <Route
      {...routeAttributes}
      render = { 
        function() {
          if(userType==='driver') {
            return element;
          }
          if(userType===null) {
            return <Redirect to="/" />
          }
          if(userType==='mom') {
            return <Redirect to="/mom/profile" />
          }
        }
      }
    />
  );
}