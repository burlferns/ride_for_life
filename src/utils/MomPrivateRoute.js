import React from 'react';
import { Route, Redirect } from "react-router-dom";
import {useSelector} from 'react-redux';

const selectorFunc = state=>state.userData.userType;

export default function(props) {
  const routeAttributes = props.routeAttributes;
  const element = props.element;
  const userType = useSelector(selectorFunc);

  return (
    <Route
      {...routeAttributes}
      render = { 
        function() {
          if(userType==='mom') {
            return element;
          }
          if(userType===null) {
            return <Redirect to="/" />
          }
          if(userType==='driver') {
            return <Redirect to="/driver/profile" />
          }
        }
      }
    />
  );
}