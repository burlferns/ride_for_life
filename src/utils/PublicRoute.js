import React from 'react';
import { Route, Redirect } from "react-router-dom";
import {useSelector} from 'react-redux';

const selectorFunc = state=>state.userData.userType;

export default function(props) {
  const routeAttributes = props.routeAttributes;
  const userType = useSelector(selectorFunc);

  return (
    <Route
      {...routeAttributes}
      render = { 
        function() {
          if(userType===null) {
            return props.element;
          }
          if(userType==='mom') {
            return <Redirect to="/mom/profile" />
          }
          if(userType==='driver') {
            return <Redirect to="/driver/profile" />
          }
        }
      }
    />

    // null

  );
}