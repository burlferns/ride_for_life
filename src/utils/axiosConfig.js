import axios from "axios";

/*******************************************************************

      !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
                        IMPORTANT NOTE
      !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

      I later learned that this way of writing the module creates
      a new axios instance whenever we use either of the functions.

      This means we create many axios instances, as we use the functions
      a lot.

      We only need 2 axios instances for this project and not so many.
      See my Docs/axios folder for more information on how to do better
      in the future.



********************************************************************/




export function axiosWithAuth() {
  const token = localStorage.getItem('authToken');

  return axios.create({
    baseURL: process.env.REACT_APP_BACKEND_URL,
    headers: {
      'Authorization': `${token}`,
    },
  });
};


export function axiosNoAuth() {
  return axios.create({
    baseURL: process.env.REACT_APP_BACKEND_URL
  });
};
