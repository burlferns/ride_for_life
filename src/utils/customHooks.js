import {useState, useEffect, useRef} from 'react';

/*
  This hook can measure the size of an element. It has to be called in the following way:
  const [elemRef,elemSize] = useSizeObserver();

  where:
    elemRef must be set to the ref attribute of the ReactDOMelement that is to be measured
    elemSize is an object that contains the width and height of the measured ReactDOMelement
    So the height is obtained as elemSize.height and the width as elemSize.width 
*/
export const useSizeObserver = function () {
  const elemRef = useRef(null);  //Reference to element that is measured
  const observerRef = useRef(null); //Reference to the resizeObserver object
  const [elemSize,setElemSize] = useState({width:0, height:0}); //This stores the measurements 
                                                                //of the element

  useEffect(()=>{
    observerRef.current = new ResizeObserver(entries=>{
      const entry = entries[0]; //The ResizeObserver can observe more than one element at a time
                                //so that is why entries is an array of measurement objects for 
                                //each of the elements that ResizeObserver observes. In this hook,
                                //since the resizeObserver is only looking at one element, we only 
                                //use the first entry in the array
      setElemSize({width:entry.contentRect.width, height:entry.contentRect.height});
    });
    observerRef.current.observe(elemRef.current);

    //I was getting a warning from React when I used elemRef.current in the unobserver function
    //that is run on dismount, saying: 
    //    "The ref value 'elemRef.current' will likely have changed by the time this effect cleanup
    //    function runs. If this ref points to a node rendered by React, copy 'elemRef.current'
    //    to a variable inside the effect, and use that variable in the cleanup function"
    //So that is why the variable refToElement is used here 
    const refToElement = elemRef.current;

    return ()=>{observerRef.current.unobserve(refToElement)};
  },[])
  
  return [elemRef,elemSize]
}