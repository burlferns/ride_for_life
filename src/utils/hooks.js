import {useState, useEffect, useRef, useContext} from 'react';

/*
  This hook can measure the size of an element. It has to be called in the following way:
  const [elemRef,elemSize] = useSizeObserver();

  where:
    elemRef must be set to the ref attribute of the ReactDOMelement that is to be measured
    elemSize is an object that contains the width and height of the measured ReactDOMelement
    So the height is obtained as elemSize.height and the width as elemSize.width 
*/
export const useSizeObserver = function () {
  const elemRef = useRef(null);  //Reference to element that is measured
  const observerRef = useRef(null); //Reference to the resizeObserver object
  const [elemSize,setElemSize] = useState({width:0, height:0}); //This stores the measurements 
                                                                //of the element

  useEffect(()=>{
    observerRef.current = new ResizeObserver(entries=>{
      const entry = entries[0]; //The ResizeObserver can observe more than one element at a time
                                //so that is why entries is an array of measurement objects for 
                                //each of the elements that ResizeObserver observes. In this hook,
                                //since the resizeObserver is only looking at one element, we only 
                                //use the first entry in the array
      setElemSize({width:entry.contentRect.width, height:entry.contentRect.height});
    });
    observerRef.current.observe(elemRef.current);

    return ()=>{observerRef.current.unobserve(elemRef.current)};
  },[])
  
  return [elemRef,elemSize]
}