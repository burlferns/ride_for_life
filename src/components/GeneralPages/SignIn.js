import React from 'react';
import styled from "styled-components";

import SignInForm from '../Form/SignInForm';
import img1 from '../../images/random-institute-JP9BQVlcED8-unsplash.jpg';

const ComponentContainer = styled.div`
  width: 100vw;
  height: calc(100vh - 8.4rem);
  min-height: 68.2rem; //60.2rem {Height of DivContainer} + 4rem*2 {min margin}
  display:grid;
  grid-template-columns: minmax(4rem, 1fr) minmax(28rem, 95.2rem) minmax(4rem, 1fr);
  grid-template-rows: minmax(4rem, 1fr) 60.2rem minmax(4rem, 1fr);
  grid-template-areas: ". . ." ". container ." ". . .";

  @media (max-width: 599px) {
    grid-template-columns: minmax(2rem, 1fr) minmax(28rem, 95.2rem) minmax(2rem, 1fr);
    grid-template-rows: minmax(2rem, 1fr) 65.1rem minmax(2rem, 1fr);
    min-height: 69.1rem; //65.1rem {Height of DivContainer} + 2rem*2 {min margin}
  }

`;

const DivContainer = styled.div`
  height:60rem;
  background:transparent;
  box-shadow:  4px 4px 3px 0px #f2f2f2;
  border:1px solid #e6e6e6;
  box-sizing: content-box;

  grid-area: container;

  display:grid;
  grid-template-columns: 1fr 34rem;
  grid-template-rows: 100%;
  grid-template-areas: "picture content";

  @media (max-width: 599px) {
    height:fit-content;

    display:grid;
    grid-template-columns: auto;
    grid-template-rows: auto;
    grid-template-areas: "picture" "content";
  }
`; 

const PictureDiv = styled.div`
  height: 60rem;
  background-image: url(${img1});
  background-size: cover;
  background-position: bottom center;

  grid-area: picture;

  @media (max-width: 599px) {
    height:20rem;
    min-width:26rem;
  }
`;


const ContentsDiv = styled.div`
  padding: 0 4rem;
  margin: auto 0;
  width:fit-content;
  box-sizing: content-box;

  grid-area: content;

  @media (max-width: 599px) {
    padding:0;
    padding-top:2rem;
    padding-bottom:2rem;
    margin:0;

    justify-self: center;
  }
`;


const StylH1 = styled.h1`
  font-size: 2.5rem;
  margin-bottom: 2.5rem;
`;

export default function SignIn() {

  return (
    <ComponentContainer>

      <DivContainer>
        <PictureDiv></PictureDiv>
        <ContentsDiv>
          <StylH1>Login to Ride for Life</StylH1>   
          <SignInForm/>
        </ContentsDiv>
      </DivContainer>

    </ComponentContainer>
    );
}