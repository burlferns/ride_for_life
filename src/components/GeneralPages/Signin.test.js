import React from 'react';
import { Provider } from 'react-redux';
import { render, screen, } from '@testing-library/react';
import configureStore from 'redux-mock-store';
import { BrowserRouter} from "react-router-dom";
import 'jest-styled-components';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';
import 'jest-localstorage-mock';
import * as axiosFunc from '../../utils/axiosConfig';

import SignIn from './SignIn.js';

import { conLog } from '../Form/SignInForm.js';

const middlewares = [];
const mockStore = configureStore(middlewares);



describe('Tests for SignIn page', () => {

  let container;
  let store;

  beforeEach(() => {
    // Initialize mockstore with empty state
    const initialState = {}
    store = mockStore(initialState)

    const renderReturn  = render(
      <Provider store={store}>
        <BrowserRouter>
          <SignIn />
        </BrowserRouter>        
      </Provider>
    )
    container = renderReturn.container;
    localStorage.clear();
  });


  test('snapshot of SignIn page', () => {
    expect(container.firstChild).toMatchSnapshot();
  })


  test('error message after clicking and leaving email field', async ()=>{
    //Error message not there initially
    const search1 = screen.queryByText('Email address is required', { selector: 'p' });
    expect(search1).not.toBeInTheDocument();

    // screen.debug();

    //Now click on email input and then click outside email input
    const titleH1 = screen.getByText('Login to Ride for Life');
    const emailInput = screen.getByTestId('email');
    userEvent.click(emailInput);
    userEvent.click(titleH1);

    //Error message should be there now
    const search2 = await screen.findByText('Email address is required', { selector: 'p' });
    expect(search2).toBeInTheDocument();

    // screen.debug();
  })


  test('error message after clicking and leaving password field', async ()=>{
    //Error message not there initially
    const search1 = screen.queryByText('Please input a password', { selector: 'p' });
    expect(search1).not.toBeInTheDocument();

    //Now click on password input and then click outside password input
    const titleH1 = screen.getByText('Login to Ride for Life');
    const passInput = screen.getByTestId('passwd');
    userEvent.click(passInput);
    userEvent.click(titleH1);

    //Error message should be there now
    const search2 = await screen.findByText('Please input a password', { selector: 'p' });
    expect(search2).toBeInTheDocument();
  })


  test('error message after clicking and leaving account type drop down', async ()=>{
    //Error message not there initially
    const search1 = screen.queryByText('Please choose account type', { selector: 'p' });
    expect(search1).not.toBeInTheDocument();

    //Now click on account type drop down and then click outside it
    const titleH1 = screen.getByText('Login to Ride for Life');
    const dropDown = screen.getByTestId('userType');
    userEvent.click(dropDown);
    userEvent.click(titleH1);

    //Error message should be there now
    const search2 = await screen.findByText('Please choose account type', { selector: 'p' });
    expect(search2).toBeInTheDocument();
  })


  test('error message to input proper email', async ()=>{
    //Error message not there initially
    const search1 = screen.queryByText('Please input a valid email address', { selector: 'p' });
    expect(search1).not.toBeInTheDocument();

    // screen.debug();

    //Now click on email input and then click outside email input
    const titleH1 = screen.getByText('Login to Ride for Life');
    const emailInput = screen.getByTestId('email');
    userEvent.click(emailInput);
    userEvent.click(titleH1);
    await userEvent.type(emailInput,'abc');
    
    //Error message should be there now
    const search2 = await screen.findByText('Please input a valid email address', { selector: 'p' });
    expect(search2).toBeInTheDocument();

    // screen.debug();

    await userEvent.type(emailInput,'abc@def');

    //Error message should still be there
    const search3 = await screen.findByText('Please input a valid email address', { selector: 'p' });
    expect(search3).toBeInTheDocument();

    // screen.debug();

    await userEvent.type(emailInput,'abc@def.com');

    //Error message should still be there
    const search4 = await screen.findByText('Please input a valid email address', { selector: 'p' });
    expect(search4).not.toBeInTheDocument();

    // screen.debug();
  })

  test('error message after clicking login button without inputs', async ()=>{
    const search1A = screen.queryByText('Email address is required', { selector: 'p' });
    const search1B = screen.queryByText('Please input a password', { selector: 'p' });
    const search1C = screen.queryByText('Please choose account type', { selector: 'p' });
    expect(search1A).not.toBeInTheDocument();
    expect(search1B).not.toBeInTheDocument();
    expect(search1C).not.toBeInTheDocument();
    
    const lgInButton = screen.getByText('Login');
    userEvent.click(lgInButton);

    const search2A = await screen.findByText('Email address is required', { selector: 'p' });
    const search2B = screen.queryByText('Please input a password', { selector: 'p' });
    const search2C = screen.queryByText('Please choose account type', { selector: 'p' });
    expect(search2A).toBeInTheDocument();
    expect(search2B).toBeInTheDocument();
    expect(search2C).toBeInTheDocument();    
  })

  test('successful mom login', async ()=>{
    //Enter mom's email and password for login
    const emailInput = screen.getByTestId('email');
    await userEvent.type(emailInput,'mom@gmail.com');
    const passInput = screen.getByTestId('passwd');
    await userEvent.type(passInput,'goodPass');

    //Select mom as the user type
    let search = screen.queryByText('Mom', { selector: 'p' });
    expect(search).not.toBeInTheDocument();
    // screen.debug();
    const dropDown = screen.getByTestId('userType');
    userEvent.click(dropDown);
    // screen.debug();
    const dropDownMom = screen.getByText('Mom', { selector: 'button' })
    userEvent.click(dropDownMom);
    search = await screen.findByText('Mom', { selector: 'p' });
    expect(search).toBeInTheDocument();
    // screen.debug();   
    
    expect(localStorage.__STORE__).toEqual({});
    search = screen.queryByText('Please Wait...', { selector: 'p' });  
    expect(search).not.toBeInTheDocument();

    //Click the login button to login a mom
    let axiosNoAuthPostData = {
      users_email: 'mom@gmail.com',
      password: 'goodPass'
    }
    let axiosNoAuthPostReturn = {
      data : {
        token: 'abcd-token',
        id: '123-id'
      }
    }
    let axiosNoAuthPostFunc = jest.fn(  ()=>axiosNoAuthPostReturn  );
    axiosFunc.axiosNoAuth = jest.fn(  ()=>({post:axiosNoAuthPostFunc})  );
    let axiosWithAuthGetReturn = {
      data : {
        users_name: 'momTest',
        users_plot: 123,
        users_phone_number: 456,
        users_email: 'mailTest' 
      }
    }
    let axiosWithAuthGetFunc = jest.fn(  ()=>axiosWithAuthGetReturn  );
    axiosFunc.axiosWithAuth = jest.fn(  ()=>({get:axiosWithAuthGetFunc})  );
    const actions = store.getActions();
    let reducerData = {
      type: 'userData/setUserData',
      payload: {
        userType: 'mom',
        users_name: 'momTest',
        users_plot: 123,
        users_phone_number: 456,
        users_email: 'mailTest'
      }
    };
    const lgInButton = screen.getByText('Login');
    userEvent.click(lgInButton);   

    search = await screen.findByText('Please Wait...', { selector: 'p' });    

    expect(axiosNoAuthPostFunc)
      .toHaveBeenCalledWith('/api/auth/user_login',axiosNoAuthPostData);
    expect(localStorage.__STORE__['authToken']).toBe('abcd-token');
    expect(localStorage.__STORE__['userId']).toBe('123-id');
    expect(axiosWithAuthGetFunc)
      .toHaveBeenCalledWith(`/api/users/123-id`);
    expect(actions).toEqual([reducerData]);
  })



  test('bad mom credentials', async ()=>{
    //Enter mom's email and password for login
    const emailInput = screen.getByTestId('email');
    await userEvent.type(emailInput,'mom@gmail.com');
    const passInput = screen.getByTestId('passwd');
    await userEvent.type(passInput,'badPass');

    //Select mom as the user type
    let search = screen.queryByText('Mom', { selector: 'p' });
    expect(search).not.toBeInTheDocument();
    
    const dropDown = screen.getByTestId('userType');
    userEvent.click(dropDown);
    
    const dropDownMom = screen.getByText('Mom', { selector: 'button' })
    userEvent.click(dropDownMom);
    search = await screen.findByText('Mom', { selector: 'p' });
    expect(search).toBeInTheDocument();
        
    expect(localStorage.__STORE__).toEqual({});
    search = screen.queryByText('Please Wait...', { selector: 'p' });  
    expect(search).not.toBeInTheDocument();

    //Click the login button but don't login mom because of bad credentials
    let axiosNoAuthPostData = {
      users_email: 'mom@gmail.com',
      password: 'badPass'
    };
    let axiosNoAuthPostReturnData = {
      response : {
        data : {
          message: 'Invalid credentials'
        }
      }
    };
    let axiosNoAuthPostReject;
    let axiosNoAuthPostReturn = new Promise((ful,rej)=>{
      axiosNoAuthPostReject = ()=>rej(axiosNoAuthPostReturnData);
    });
    let axiosNoAuthPostFunc = jest.fn(  ()=>axiosNoAuthPostReturn  );
    axiosFunc.axiosNoAuth = jest.fn(  ()=>({post:axiosNoAuthPostFunc})  );

    const lgInButton = screen.getByText('Login');
    userEvent.click(lgInButton);   

    search = await screen.findByText('Please Wait...', { selector: 'p' });  

    expect(axiosNoAuthPostFunc)
      .toHaveBeenCalledWith('/api/auth/user_login',axiosNoAuthPostData);
    expect(axiosNoAuthPostFunc).toHaveReturnedWith(axiosNoAuthPostReturn);

    search = screen.queryByText(
      'Incorrect login credentials used, please try again', 
      { selector: 'p' }
    );  
    expect(search).not.toBeInTheDocument();

    //Make the noAuth post API request return an error from server 
    axiosNoAuthPostReject();
    
    search = await screen.findByText(
      'Incorrect login credentials used, please try again', 
      { selector: 'p' }
    );  

    expect(localStorage.__STORE__).toEqual({});
    search = screen.queryByText('Please Wait...', { selector: 'p' });  
    expect(search).not.toBeInTheDocument();
  })


  //Tests for successful driver login and unsuccessful driver login 
  //are not here in this demo project, because they are just copy and
  //paste of the last two tests with some modifications.


});