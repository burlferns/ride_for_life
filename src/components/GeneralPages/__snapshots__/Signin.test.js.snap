// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for SignIn page snapshot of SignIn page 1`] = `
.c6 {
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  display: grid;
  grid-template-columns: 4rem 22rem;
  grid-template-rows: 4rem auto;
  grid-template-areas: "divImg divInput" "divError divError";
}

.c7 {
  width: 4rem;
  height: 4rem;
  border-top-left-radius: 0.5rem;
  border-bottom-left-radius: 0.5rem;
  background: #E5E7E9;
  border: 0.1rem solid #A6ACAF;
  grid-area: divImg;
}

.c8 {
  width: 1.6rem;
  height: 1.6rem;
  margin-top: 1.2rem;
  margin-left: 1.2rem;
}

.c9 {
  width: 22rem;
  height: 4rem;
  border-top-right-radius: 0.5rem;
  border-bottom-right-radius: 0.5rem;
  border: 0.1rem solid #A6ACAF;
  position: relative;
  top: 0;
  left: 0;
  grid-area: divInput;
}

.c9:focus-within::after {
  content: '';
  position: absolute;
  box-sizing: border-box;
  top: -0.2rem;
  left: -0.2rem;
  bottom: -0.2rem;
  right: -0.2rem;
  border-top-right-radius: 0.5rem;
  border-bottom-right-radius: 0.5rem;
  border: 0.2rem solid #5DADE2;
}

.c10 {
  width: 20rem;
  margin-top: 1.1rem;
  margin-left: 1rem;
  font-size: 1.6rem;
  color: #5F6A6A;
  border: none;
  outline: none;
  background: transparent;
}

.c11 {
  width: 20rem;
  margin-top: 1.1rem;
  margin-left: 1rem;
  font-size: 1.6rem;
  color: #5F6A6A;
  position: absolute;
  top: 0;
  left: 0;
  z-index: -1;
}

.c12 {
  width: 26rem;
  height: 2.9rem;
  margin-top: 0.5rem;
  grid-area: divError;
}

.c13 {
  width: 26rem;
  height: 4rem;
  border-radius: 0.5rem;
  border: 0.1rem solid #A6ACAF;
  position: relative;
  top: 0;
  left: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  outline: none;
}

.c13:focus::after {
  content: '';
  position: absolute;
  box-sizing: border-box;
  top: -0.2rem;
  left: -0.2rem;
  bottom: -0.2rem;
  right: -0.2rem;
  border-radius: 0.5rem;
  border: 0.2rem solid #5DADE2;
}

.c14 {
  color: #5F6A6A;
  font-size: 1.6rem;
  margin-top: 1.1rem;
  margin-left: 1.5rem;
}

.c15 {
  width: 1rem;
  height: 1rem;
  position: absolute;
  top: 1.4rem;
  right: 0.2rem;
}

.c16 {
  width: 26rem;
  margin-top: 0.5rem;
  height: 2.9rem;
}

.c17 {
  width: 26rem;
  height: 4rem;
  border-radius: 0.5rem;
  position: relative;
  top: 0;
  left: 0;
}

.c18 {
  width: 26rem;
  height: 4rem;
  border-radius: 0.5rem;
  font-size: 1.6rem;
  color: white;
  background: #3C97D7;
  outline: none;
  border: none;
}

.c18:active {
  border: 3px solid white;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
}

.c19 {
  margin-top: 0.5rem;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  height: 1.2rem;
}

.c20 {
  width: 26rem;
  margin-top: 3rem;
}

.c22 {
  width: 25rem;
  height: 2rem;
  background: transparent;
}

.c21 {
  font-size: 1.2rem;
}

.c0 {
  width: 100vw;
  height: calc(100vh - 8.4rem);
  min-height: 68.2rem;
  display: grid;
  grid-template-columns: minmax(4rem,1fr) minmax(28rem,95.2rem) minmax(4rem,1fr);
  grid-template-rows: minmax(4rem,1fr) 60.2rem minmax(4rem,1fr);
  grid-template-areas: ". . ." ". container ." ". . .";
}

.c1 {
  height: 60rem;
  background: transparent;
  box-shadow: 4px 4px 3px 0px #f2f2f2;
  border: 1px solid #e6e6e6;
  box-sizing: content-box;
  grid-area: container;
  display: grid;
  grid-template-columns: 1fr 34rem;
  grid-template-rows: 100%;
  grid-template-areas: "picture content";
}

.c2 {
  height: 60rem;
  background-image: url(random-institute-JP9BQVlcED8-unsplash.jpg);
  background-size: cover;
  background-position: bottom center;
  grid-area: picture;
}

.c3 {
  padding: 0 4rem;
  margin: auto 0;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  box-sizing: content-box;
  grid-area: content;
}

.c4 {
  font-size: 2.5rem;
  margin-bottom: 2.5rem;
}

@media (max-width:599px) {
  .c0 {
    grid-template-columns: minmax(2rem,1fr) minmax(28rem,95.2rem) minmax(2rem,1fr);
    grid-template-rows: minmax(2rem,1fr) 65.1rem minmax(2rem,1fr);
    min-height: 69.1rem;
  }
}

@media (max-width:599px) {
  .c1 {
    height: -webkit-fit-content;
    height: -moz-fit-content;
    height: fit-content;
    display: grid;
    grid-template-columns: auto;
    grid-template-rows: auto;
    grid-template-areas: "picture" "content";
  }
}

@media (max-width:599px) {
  .c2 {
    height: 20rem;
    min-width: 26rem;
  }
}

@media (max-width:599px) {
  .c3 {
    padding: 0;
    padding-top: 2rem;
    padding-bottom: 2rem;
    margin: 0;
    justify-self: center;
  }
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <div
      class="c2"
    />
    <div
      class="c3"
    >
      <h1
        class="c4"
      >
        Login to Ride for Life
      </h1>
      <form
        class="c5"
      >
        <div
          class="c6"
        >
          <div
            class="c7"
          >
            <img
              class="c8"
              src="envelope.svg"
            />
          </div>
          <div
            class="c9"
          >
            <input
              class="c10"
              data-testid="email"
              name="email"
              type="email"
              value=""
            />
            <p
              class="c11"
            >
              Email Address
            </p>
          </div>
          <div
            class="c12"
            height="2.9rem"
          />
        </div>
        <div
          class="c6"
        >
          <div
            class="c7"
          >
            <img
              class="c8"
              src="lock.svg"
            />
          </div>
          <div
            class="c9"
          >
            <input
              class="c10"
              data-testid="passwd"
              name="passwd"
              type="password"
              value=""
            />
            <p
              class="c11"
            >
              Password
            </p>
          </div>
          <div
            class="c12"
            height="2.9rem"
          />
        </div>
        <div>
          <div
            class="c13"
            data-testid="userType"
            id="userType"
            tabindex="-1"
          >
            <p
              class="c14"
            >
              Account Type
            </p>
            <img
              class="c15"
              src="drop-down.svg"
            />
          </div>
          <div
            class="c16"
            height="2.9rem"
          />
        </div>
        <div
          class="c17"
        >
          <button
            class="c18"
            type="submit"
          >
            Login
          </button>
        </div>
        <div
          class="c19"
        />
        <div
          class="c20"
        >
          <a
            class="c21"
            href="/mom/register"
          >
            Are you an expecting mother? Register here.
          </a>
          <div
            class="c22"
          />
          <a
            class="c21"
            href="/driver/register"
          >
            Want to be a driver? Register here.
          </a>
        </div>
      </form>
    </div>
  </div>
</div>
`;
